aikido_perception/src/AprilTagsModule.cpp

```diff
@@ -0,0 +1,187 @@
+void AprilTagsModule::detectObjects(std::vector<dart::dynamics::SkeletonPtr>& skeleton_list,double _timeout, ros::Time timestamp)
+{
+	//Looks at all detected tags, looks up config file 
+	//Appends new skeletons to skeleton list
+	visualization_msgs::MarkerArrayConstPtr marker_message
+ 			= ros::topic::waitForMessage<visualization_msgs::MarkerArray>(mMarkerTopic,mNode,ros::Duration(_timeout));
+
+	for(size_t i=0; i < marker_message->markers.size(); i++)
+	{
+		//Get marker, tag ID, and detection transform name 
+		visualization_msgs::Marker marker_transform = marker_message->markers[i];
+		ros::Time marker_stamp = marker_transform.header.stamp;
+		std::string tag_name = marker_transform.ns;
+		std::string detection_frame(marker_transform.header.frame_id);
+
+		if(marker_stamp < timestamp){
+			dtwarn << "Marker stamp " << marker_stamp<< " for " << tag_name 
+			<< "is behind timestamp parameter: " <<timestamp <<"\n";
+
+			continue;
+		}
+
+		std::string body_name;
+		Eigen::Matrix4d skel_offset;
+
+		//check if tag name is in database
+		if (mConfigData->getTagNameOffset(tag_name,body_name,skel_offset))
+		{
+			//Get orientation of marker
+			Eigen::Quaterniond orien(marker_transform.pose.orientation.w,
+									marker_transform.pose.orientation.x,
+									marker_transform.pose.orientation.y,
+									marker_transform.pose.orientation.z);
+
+			//Get in 4x4 pose form
+			Eigen::Vector3d qpos(marker_transform.pose.position.x,marker_transform.pose.position.y,marker_transform.pose.position.z);
+			Eigen::Matrix3d qrot(orien.toRotationMatrix());
+			Eigen::Matrix4d marker_pose;
+			marker_pose<< qrot,qpos,0,0,0,1;
+
+			//For the frame-frame transform
+			tf::StampedTransform transform;
+			tf::TransformListener listener;
+
+			try{
+				listener.waitForTransform(mDestinationFrame,detection_frame,
+					ros::Time(0),ros::Duration(_timeout));
+
+				listener.lookupTransform(mDestinationFrame,detection_frame,
+					ros::Time(0), transform);
+			}
+			catch(tf::TransformException ex){
+				ROS_ERROR("%s",ex.what());
+			}
+
+			//Get frame pose
+			tf::Quaternion tf_quat(transform.getRotation());
+
+			Eigen::Quaterniond frame_quat(tf_quat.getW(),tf_quat.getX(),tf_quat.getY(),tf_quat.getZ());
+			Eigen::Vector3d frame_trans(transform.getOrigin());
+			Eigen::Matrix3d frame_rot(frame_quat.toRotationMatrix());
+			Eigen::Matrix4d frame_pose;
+			frame_pose<<frame_rot,frame_trans,0,0,0,1;
+
+			//Compose to get actual skeleton pose
+			Eigen::Isometry3d temp_pose;
+			temp_pose.matrix() = frame_pose * marker_pose;
+			Eigen::Isometry3d skel_pose;
+			skel_pose.matrix() = temp_pose * skel_offset;
+			Eigen::Isometry3d link_offset = mReferenceLink->getWorldTransform();
+			skel_pose = link_offset * skel_pose;
+
+			//Check if skel in skel_list
+			//If there then update pose
+			//If not then append skeleton
+			std::string skel_name = body_name.substr(0,body_name.find('.'));
```

Comment:
Changed this so that the `YAML` file has an additional `resource` field for each tag, which points to the URDF file of the model. The `name` field now just has the object name. This obviates the need for 1) A URDF folder path  and 2) Deriving the object name from the filename.